sonar-helm-for-iac/converters/tree_converter_test.go

```diff
@@ -34,44 +50,124 @@ func Test_convert_simple(t *testing.T) {
 // This test describes the existing behavior until we need to change it.
 func Test_comments_are_ignored(t *testing.T) {
-	tpl, _ := template.New("test").Parse("{{/* comment */}}")
-	node := convert(tpl.Root).(*pbstructs.ListNode)
+	code := "{{/* comment */}}"
+	tpl, _ := template.New("test").Parse(code)
+	ctx := ConversionContext{
+		Content:   code,
+		Converter: defaultTestConverter,
+	}
+	node := ctx.Convert(tpl.Root).(*pbstructs.ListNode)
 	assert.Equal(t, 0, len(node.Nodes))
 }
 
 // This test shows how to enable comments in the AST.
 func Test_convert_comments(t *testing.T) {
+	code := "{{/* comment */}}"
 	tree := parse.New("test")
 	tree.Mode = parse.ParseComments
-	tree, _ = tree.Parse("{{/* comment */}}", "", "", make(map[string]*parse.Tree))
-	node := convert(tree.Root).(*pbstructs.ListNode)
+	tree, _ = tree.Parse(code, "", "", make(map[string]*parse.Tree))
+	ctx := ConversionContext{
+		Content:   code,
+		Converter: defaultTestConverter,
+	}
+
+	node := ctx.Convert(tree.Root).(*pbstructs.ListNode)
+
 	assert.Equal(t, 1, len(node.Nodes))
 	n1, _ := anypb.UnmarshalNew(node.Nodes[0], proto.UnmarshalOptions{})
 	assert.True(t, strings.HasSuffix(node.Nodes[0].TypeUrl, "CommentNode"))
 	commentNode := n1.(*pbstructs.CommentNode)
 	assert.Equal(t, "/* comment */", *commentNode.Text)
 }
 
-func Test_TreeConvert(t *testing.T) {
-	tests := []struct {
-		name string
-		tmpl string
-	}{
-		{
-			name: "Test_TreeConvert_simple",
-			tmpl: "{{ . }}",
-		},
-		{
-			name: "Test_TreeConvert_define_and_template",
-			tmpl: `
+func Test_TreeConvert_simple_dot(t *testing.T) {
+	code := "{{ . }}"
+	tpl, _ := template.New("test-simple-dot").Parse(code)
+
+	tree := defaultTestConverter.ConvertTree(code, tpl.Tree)
+
+	basicTreeAsserts(t, tree, "test-simple-dot", 5, 0, 1)
+	n, _ := anypb.UnmarshalNew(tree.Root.Nodes[0], proto.UnmarshalOptions{})
+	actionNode := n.(*pbstructs.ActionNode)
+	assert.Equal(t, int64(3), actionNode.Pos)
+	assert.Equal(t, int64(5), actionNode.Length)
+	assert.Equal(t, int64(1), actionNode.Pipe.Length)
+	assert.Empty(t, actionNode.Pipe.Decl)
+	assert.Equal(t, 1, len(actionNode.Pipe.Cmds))
+	cmd := actionNode.Pipe.Cmds[0]
+	assert.Equal(t, int64(1), cmd.Length)
+	assert.Equal(t, 1, len(cmd.Args))
+	n, _ = anypb.UnmarshalNew(cmd.Args[0], proto.UnmarshalOptions{})
+	dot := n.(*pbstructs.DotNode)
+	assert.Equal(t, int64(3), dot.Pos)
+	assert.Equal(t, int64(1), dot.Length)
+}
+
+func Test_TreeConvert_define_and_template(t *testing.T) {
+	code := `
 {{- /* returns "foo" */ -}}
 {{- define "foo" }}
 {{ slice 2 "00foo" }}
 {{ end }}
 
 {{ template "foo" }}
-`,
-		},
+`
+	tpl, _ := template.New("test-define-and-template").Parse(code)
+
+	tree := defaultTestConverter.ConvertTree(code, tpl.Tree)
+
+	basicTreeAsserts(t, tree, "test-define-and-template", 21, 29, 3)
+
+	n, _ := anypb.UnmarshalNew(tree.Root.Nodes[0], proto.UnmarshalOptions{})
+	textNode := n.(*pbstructs.TextNode)
+	// Note: comment and `define` nodes are not present in the AST
```

Comment:
Is there an option to include `define` nodes to AST? This will be needed for raising issues in custom templates.

Reply:
Not easily. It is parsed and added into `Tree.treeSet`, so that it can be later used during execution. It can be extracted from there and placed into the tree (however, there is no node type for `define`). But maybe it's another point in favor of implementing a our own parser.